katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+    """An error occurred in loading calibration values from file"""
+    pass
+
+
+class CalibrationValueError(ValueError):
+    """An error occurred in the calculation of calibration coefficients"""
+    pass
+
+
+class SimpleInterpolate1D(object):
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+
+    def __call__(self, x):
+        y_re = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        y_im = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        # Interpolate across nans in time axis.
+        for ch_idx in range(self._y.shape[1]):
+            for pol_idx in range(self._y.shape[2]):
+                for ant_idx in range(self._y.shape[3]):
+                    y_re[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                            self._y[:, ch_idx, pol_idx, ant_idx].real,
+                                                                            x)
+                    y_im[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                            self._y[:, ch_idx, pol_idx, ant_idx].imag,
+                                                                            x)
+        y = y_re + y_im
+        return y
+
+
+class ComplexInterpolate1D(object):
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+        self._mag = np.abs(y)
+        self._phase = y / self._mag
+
+    def __call__(self, x):
+        mag = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        phase_re = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        phase_im = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        # Interpolate across nans in time axis.
+        for ch_idx in range(self._y.shape[1]):
+            for pol_idx in range(self._y.shape[2]):
+                for ant_idx in range(self._y.shape[3]):
+                    mag[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                           self._mag[:, ch_idx, pol_idx, ant_idx],
+                                                                           x)
+                    phase_re[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                                self._phase[:, ch_idx, pol_idx, ant_idx].real,
+                                                                                x)
+                    phase_im[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                                self._phase[:, ch_idx, pol_idx, ant_idx].imag,
+                                                                                x)
+        phase = phase_re + phase_im
```

Comment:
Presumably the phases could be interpolated by calling SimpleInterpolate1D, instead of duplicating the logic to split into re and im.